db/compaction/compaction_service_test.cc

```diff
@@ -227,6 +246,189 @@ class CompactionServiceTest : public DBTestBase {
+    }
+    auto cro = CompactRangeOptions();
+    if (compaction_has_subcompaction) {
+      cro.max_subcompactions = 10;
+    }
+    Status s = db_->CompactRange(cro, nullptr, nullptr);
+    if (compaction_unfinished) {
+      ASSERT_NOK(s);
+      my_cs->ResetOverride();
+    } else {
+      ASSERT_OK(s);
+    }
+
+    SyncPoint::GetInstance()->ClearAllCallBacks();
+    SyncPoint::GetInstance()->DisableProcessing();
+  }
+
+  void ReopenDB(
+      bool resume_compaction = true,
+      bool resume_compaction_concurrent_with_manual_compaction = false) {
+    std::vector<LiveFileMetaData> metadata;
+    db_->GetLiveFilesMetaData(&metadata);
+    auto prev_reopen_live_file_num = metadata.size();
+
+    SyncPoint::GetInstance()->SetCallBack(
+        "DBImpl::NotifyOnCompactionBegin::UnlockMutex", [&](void*) {
+          if (resume_compaction_concurrent_with_manual_compaction) {
+            std::vector<std::string> input_files;
+            for (const auto& file : metadata) {
+              input_files.push_back(file.name);
+            }
+            ASSERT_EQ(input_files.size(), 2);
+            Status s = db_->CompactFiles(CompactionOptions(), input_files, 1);
+            ASSERT_TRUE(s.IsAborted());
+          } else if (compaction_has_penultimate_level_) {
+            auto* cfd = static_cast_with_check<ColumnFamilyHandleImpl>(
+                            db_->DefaultColumnFamily())
+                            ->cfd();
+
+            auto* f = new FileMetaData();
+            std::string key = "k1";
+            f->smallest = InternalKey(Slice(key), 4 /* seqno */, kTypeValue);
+            f->largest = InternalKey(Slice(key), 4 /* seqno */, kTypeValue);
+
+            std::vector<CompactionInputFiles> inputs;
+            inputs.push_back(CompactionInputFiles());
+            inputs[0].level = 0;
+            inputs[0].files.push_back(f);
+
+            // Verify this compaction with output level L1 is overlapped with
+            // the resumable compaction with penultimate output level L1
+            ASSERT_TRUE(
+                cfd->compaction_picker()->FilesRangeOverlapWithCompaction(
+                    inputs, 1 /* output_level */, -1 /* penultimate level*/));
+            delete f;
+          }
+        });
+    SyncPoint::GetInstance()->EnableProcessing();
+
+    options_.resume_compaction = resume_compaction;
+    if (resume_compaction) {
+      options_.disable_auto_compactions = false;
+    }
+    DummyEventListner* listener = new DummyEventListner();
+    options_.listeners.emplace_back(listener);
+
+    Reopen(options_);
+    ASSERT_OK(dbfull()->TEST_WaitForCompact());
+
+    SyncPoint::GetInstance()->ClearAllCallBacks();
+    SyncPoint::GetInstance()->DisableProcessing();
+
+    auto my_cs = GetCompactionService();
+
+    metadata.clear();
+    db_->GetLiveFilesMetaData(&metadata);
+    if (compaction_unfinished_ && resume_compaction) {
+      ASSERT_LT(metadata.size(), prev_reopen_live_file_num);
```

Comment:
TODO: assert sync point is called even manually tracing through debugger shows it is called.